### ISSUE TITLE
Bug: [ban-types] imported Symbol type is mistaken for the global Symbol type

### ISSUE THREAD

**Issue by jakebailey** [typescript-eslint/typescript-eslint#7306]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&fileType=.tsx&code=JYWwDg9gTgLgBAbzgZQJ4gEYQDZwL5wBmUEIcARAHQD0MqYApgM7kDcAsAFBcMAeksIgFcAdgGMYwCCLgZsAQwAWACiYAuFOizYAlIi5w4UBjCFQZTDpzxA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AIwEMnajTUMkRNGgB7aJHBgAviGlA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
import { Symbol } from "./types";

export function blah(s: Symbol) {
  return s;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/ban-types": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error about the imported `Symbol` type.

### Actual Result

The imported `Symbol` type is banned, mistaken for the global Symbol type.

**Comment by bradzacher** [comment]
This is working as expected - the `ban-types` rule is intentionally very simple in its design and just bans the type name altogether. It is contextless so that it can ban type names. This means you can do things like ban "foo" because your codebase has changed its naming convention.

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/5505, https://github.com/typescript-eslint/typescript-eslint/issues/4794, https://github.com/typescript-eslint/typescript-eslint/issues/3625
In the vast majority of cases it is what the _average_ user expects as well because they do not alias well-known global (nor do they want to!). Which is why we haven't changed it or added options. The rule itself has existed for some 6 years or something with this logic - which is why I say it's "expected".

So the question is - do we add an option to allow _local_ type names? I think there is value in this - limited as it may be. As with https://github.com/typescript-eslint/typescript-eslint/issues/6262 there is a rare case where the autofixer can break code - so we do need to add some scope analysis to the rule..

If you'd like to improve the rule with scope analysis and add an option to the type matcher that allows specific types to be configured to "local only" then happy to accept a PR! That would allow you to "extend" the defaults and make specific types local-only within your project.
Otherwise I'd consider removing `Symbol` from your config.

**Comment by jakebailey** [comment]
I think the reason I reported this was because I tried it with a type declared locally and didn't see an error; I thought that maybe it was supposed to be smarter but wasn't smarter for imported names. But trying again, it also errors when I have a local declaration with name `Symbol`, so I don't quite know what I did wrong in my first test!

How does eslint's `no-restricted-globals` work, then? It's scope-aware, so I had thought `ban-types` could be as well: https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgeCA9IHNldFRpbWVvdXQoKVxuXG5mdW5jdGlvbiBidG9hKCkge1xuICBcbn1cblxuY29uc3QgeSA9IGJ0b2EoKTsiLCJvcHRpb25zIjp7ImVudiI6eyJlczYiOnRydWV9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCJ9LCJydWxlcyI6eyJuby1yZXN0cmljdGVkLWdsb2JhbHMiOlsiZXJyb3IiLCJzZXRUaW1lb3V0IiwiYnRvYSJdfX19

**Comment by bradzacher** [comment]
Originally `no-restricted-globals` did not support types because back then we didn't have a proper scope analyser to create variables for it to understand. It would work now though I believe.

But also people have some simple usecases for types (like "don't use the `String` type") that aren't a good fit for `no-restricted-globals` because it would also ban the `String` function for coercion.

`ban-types` won't error on an interface/type declaration or on an imported name because its specifically looking for a type reference ast node.

---

I don't think the rule should use scope analysis by default - but I think it should definitely be possible to opt-in specific bans to be global-only.

**Comment by jakebailey** [comment]
> `ban-types` won't error on an interface/type declaration or on an imported name because its specifically looking for a type reference ast node.

To be clear, I meant that I wrote:

```ts
export interface Symbol {
   foo: number;
}

export function blah(s: Symbol) {
    return s;
}
```

And then didn't see an error on the use of `Symbol`... but then I retested after you replied and it errored, matching what you said, so I just tested it wrong.


> Originally `no-restricted-globals` did not support types because back then we didn't have a proper scope analyser to create variables for it to understand. It would work now though I believe.
> 
> But also people have some simple usecases for types (like "don't use the `String` type") that aren't a good fit for `no-restricted-globals` because it would also ban the `String` function for coercion.


I also wasn't saying that I could use `no-restricted-globals`; definitely that'd be bad to use on `Symbol` given you need to use that value for lots of stuff; I just meant that if eslint can do the scope analysis to figure out that it's an imported or local `Symbol`, that types could probably do that too.

**Comment by gregjacobs** [comment]
+1 for this. We import the `Symbol` type from ts-morph (a typescript compiler wrapper API) and typescript-eslint is "fixing" usages of this `Symbol` type with `symbol` (lowercase 's')

(The term "Symbol" is highly used in compilers btw. Think "Symbol Table")

Minimal example:

```ts
import { Symbol } from 'ts-morph';

export function getSymbol(): Symbol {  // <-- 'Symbol' is changed to 'symbol' here
    // ...
}
```

Would absolutely say that the `ban-types` rule should be scope-aware as mentioned in the above comment

**Comment by Josh-Cena** [comment]
I am going to close this as a duplicate of #6262. We are splitting the ban of these uppercase types into a separate more targeted rule (https://github.com/typescript-eslint/typescript-eslint/pull/9102), in which case it would be very fair for us to introduce scope analysis.
